actions/ulog_ingestion/src/ulog_ingestion/utils.py

```diff
@@ -0,0 +1,324 @@
+                    data_type=field.type_str,
+                    canonical_data_type=canonical_data_type,
+                )
+            )
+
+    return
+
+
+def create_per_topic_mcap_from_ulog(
+    output_path_per_topic_mcap: Any, d: str, json_schema_topic: Any
+) -> None:
+    """
+    Creates a per-topic MCAP file from a ULog object.
+
+    Args:
+    - output_path_per_topic_mcap: The path to the output MCAP file.
+    - d: The ULog object containing the data to be converted.
+    - json_schema_topic: The json schema dict for the topic.
+
+    Returns:
+    - None
+    """
+
+    with open(output_path_per_topic_mcap, "wb") as stream:
+        writer = Writer(stream)
+        writer.start()
+
+        schema_id = writer.register_schema(
+            name=d.name,
+            encoding="jsonschema",
+            data=json.dumps(json_schema_topic).encode(),
+        )
+
+        channel_id = writer.register_channel(
+            schema_id=schema_id,
+            topic=d.name,
+            message_encoding="json",
+        )
+
+        for i in range(len(d.data["timestamp"])):
+            values = list()
+            for f in d.field_data:
+                values.append((f.field_name, f.type_str, d.data[f.field_name][i]))
+            json_msg_instance = parse_values_to_json(values)
+
+            timestamp_ns = int(d.data["timestamp"][i] * 1000)
+
+            writer.add_message(
+                channel_id=channel_id,
+                log_time=timestamp_ns,
+                data=json.dumps(json_msg_instance).encode("utf-8"),
+                publish_time=timestamp_ns,
+            )
+
+        writer.finish()
+    return
+
+
+def setup_output_folder_structure(
+    ulog_file_path: str, input_dir: str
+) -> Tuple[str, str]:
+    """
+    Set up the output folder structure for the .mcap files.
+
+    Args:
+    - ulog_file_path: Path to the .ulg file.
+    - input_dir: Path to the input directory.
+    """
+    relative_folder_path_of_file = os.path.split(ulog_file_path.split(input_dir)[1])[0]
+
+    ulog_file_name = os.path.split(ulog_file_path)[1]
+
+    output_folder_name_ulog = ulog_file_name.replace(".ulg", "")
+    relative_folder_path_of_file = relative_folder_path_of_file.lstrip("/")
+    temp_dir = str(tempfile.TemporaryDirectory().name)
+
+    output_folder_path = os.path.join(
+        temp_dir,
+        ".VISUALIZATION_ASSETS",
+        relative_folder_path_of_file,
+        output_folder_name_ulog,
+    )
+
+    print(f"Output folder path: {output_folder_path}")
+    os.makedirs(output_folder_path, exist_ok=True)
+
+    return output_folder_path, temp_dir
+
+
+def is_valid_ulog(ulog_file_path: str) -> bool:
+    """
+    Check if the given file is a valid .ulg file.
+
+    Args:
+    - ulog_file_path: Path to the .ulg file.
+
+    Returns:
+    - True if the file is a valid .ulg file, False otherwise.
+    """
+
+    header_bytes = b'\x55\x4c\x6f\x67\x01\x12\x35'
+    file_handle = open(ulog_file_path, "rb")
```

Comment:
Minor and probably doesn't amount to much here, but, always best to `with open(...) as fd:`